src/java/htsjdk/samtools/util/BlockCompressedInputStream.java

```diff
@@ -507,6 +542,117 @@ private static boolean preambleEqual(final byte[] preamble, final byte[] buf, fi
+         */
+        private final long mBlockAddress;
+        /**
+         * Exception thrown (if any) when attempting to decompress block
+         */
+        private final Exception mException;
+        public DecompressedBlock(long blockAddress, byte[] block, int compressedSize) {
+            mBlock = block;
+            mBlockAddress = blockAddress;
+            mBlockCompressedSize = compressedSize;
+            mException = null;
+        }
+        public DecompressedBlock(long blockAddress, int compressedSize, Exception exception) {
+            mBlock = new byte[0];
+            mBlockAddress = blockAddress;
+            mBlockCompressedSize = compressedSize;
+            mException = exception;
+        }
+    }
+    
+    private void startDecompressionThread() {
+        if (mDecompressionThread == null) {
+            mDecompressionRunnable = new DecompressionRunnable();
+            mDecompressionThread = new Thread(mDecompressionRunnable, "DecompressBGZF");
+            mDecompressionThread.setDaemon(true);
+            mDecompressionThread.start();
+        }
+        if (mDecompressionRunnable.mResult == null) {
+            mDecompressionRunnable.mResult = new ArrayBlockingQueue<DecompressedBlock>(READ_AHEAD_BUFFERS);
+            assert(mDecompressionRunnable.mOperations.isEmpty());
+            for (int i = 0; i < READ_AHEAD_BUFFERS; i++) {
+                mDecompressionRunnable.mOperations.add(this);
+            }
+        }
+    }
+    
+    private void terminateDecompressionThread() {
+        if (mDecompressionThread != null) {
+            try {
+                mDecompressionRunnable.mTerminate.set(true);
+                mDecompressionRunnable.mOperations.clear();
+                mDecompressionThread.interrupt();
+                try {
+                    mDecompressionThread.join();
+                } catch (InterruptedException e) {
+                    Thread.currentThread().interrupt();
+                }                
+            } finally {
+                mDecompressionThread = null;
+                mDecompressionRunnable = null;
+            }
+        }
+    }
+    
+    private void flushDecompressionThread() {
+        if (mDecompressionThread != null) {
+            mDecompressionRunnable.mResult = null;
+            mDecompressionRunnable.mOperations.clear();
+            mDecompressionThread.interrupt();
+        }
+    }
+    
+    /**
+     * Performs block decompression
+     */
+    private class DecompressionRunnable implements Runnable {
+        /**
+         * Number of block to eagerly decompress
+         */
+        /**
+         * Each entry in this queue indicates a block to be decompressed. 
+         */
+        private BlockingQueue<Object> mOperations = new ArrayBlockingQueue<Object>(READ_AHEAD_BUFFERS);
+        /**
+         * Results of block decompression.
+         */
+        private volatile BlockingQueue<DecompressedBlock> mResult = null;
```

Comment:
Rather than updating `mResult` on seek, it might be safer and cleaner to create a new `Runnable` (this would not mean a new thread if you used a fixed size thread pool created from `Executors`). Then the `BlockingQueue` could be `final`.